src/performance/metrics/StatisticCollector.cpp

```diff
@@ -46,64 +48,172 @@ int StatisticCollector::getMemoryUsageByProcess() {
-        percent = -1.0;
-    } else {
-        percent = (timeSample.tms_stime - lastSysCPU) + (timeSample.tms_utime - lastUserCPU);
-        percent /= (now - lastCPU);
-        percent /= numProcessors;
-        percent *= 100;
+static long getSwapSpace(int field) {
+    FILE *file = fopen("/proc/swaps", "r");
+    long result = -1;
+    char line[128];
+
+    fgets(line, 128, file);
+
+    while (fgets(line, 128, file) != NULL) {
+        char *value;
+        char *save = NULL;
+        for (int i = 0; i < field; i++) {
+            if (i == 0) {
+                value = strtok_r(line, " ", &save);
+            } else {
+                value = strtok_r(NULL, "\t", &save);
+            }
+        }
+        long used = strtol(value, NULL, 10);
+        if (used < 0 || used > 0xfffffffffffffffL) {
+            continue;
+        }
+        if (result >= 0) {
+            result += used;
+        } else {
+            result = used;
+        }
     }
-    lastCPU = now;
-    lastSysCPU = timeSample.tms_stime;
-    lastUserCPU = timeSample.tms_utime;
+    fclose(file);
 
-    return percent;
+    return result;
 }
 
-std::string StatisticCollector::collectVMStatistics(std::string isVMStatManager,
-                                                    std::string isTotalAllocationRequired) {
-    std::string vmLevelStatistics;
+long StatisticCollector::getUsedSwapSpace() {
+    long result = getSwapSpace(4);
+    // std::cout << "Used swap space: " + std::to_string(result) << std::endl;
+    return result;
+}
 
-    if (isVMStatManager == "true") {
-        long totalMemoryUsed = getTotalMemoryUsage();
-        double totalCPUUsage = getTotalCpuUsage();
+long StatisticCollector::getTotalSwapSpace() {
+    long result = getSwapSpace(3);
+    // std::cout << "Total swap space: " + std::to_string(result) << std::endl;
+    return result;
+}
 
-        std::stringstream stream;
-        stream << std::fixed << std::setprecision(2) << totalCPUUsage;
-        std::string cpuUsageString = stream.str();
+long StatisticCollector::getRXBytes() {
+    FILE *file = fopen("/sys/class/net/eth0/statistics/rx_bytes", "r");
+    long result = -1;
+    fscanf(file, "%li", &result);
+    fclose(file);
+    // std::cout << "Total read bytes: " + std::to_string(result) << std::endl;
+    return result;
+}
+
+long StatisticCollector::getTXBytes() {
+    FILE *file = fopen("/sys/class/net/eth0/statistics/tx_bytes", "r");
+    long result = -1;
+    fscanf(file, "%li", &result);
+    fclose(file);
+    // std::cout << "Total sent bytes: " + std::to_string(result) << std::endl;
+    return result;
+}
 
-        vmLevelStatistics = std::to_string(totalMemoryUsed) + "," + cpuUsageString + ",";
+int StatisticCollector::getSocketCount() {
+    DIR *d = opendir("/proc/self/fd");
+    if (!d) {
+        puts("Error opening directory /proc/self/fd");
+        return -1;
+    }
+    const struct dirent *dir;
+    char path[64];
+    char link_buf[1024];
```

Comment:
Can we move 1024 to a constant. I see multiple uses of that value below.

Reply:
Those occurrences of 1024 are independent and doesn't have to be equal.

Reply:
Ok